### DISCUSSION TITLE
Any need or way to explicitly release resources from loadPage

### DISCUSSION THREAD

**Discussion by petertennis** [pymupdf/PyMuPDF#975] [Q&A] [answered]
First of all, thanks to all involved in maintaining this very useful library!

Second, please excuse a naive question but I am paranoid given some performance/memory challenges I am dealing with. (I am processing large format PDF pages in documents with a large page count ( 600+))

If I call loadPage in a for loop assigning to a single variable, is there any need to make an explicit call to release page resources at the end of the loop? Or will this happen automatically as the reference to the old object is overwritten each time around?

document = fitz.open(local_document_filename)
for page_num in range(document.pageCount):
	page = document.loadPage(page_num)
	#do stuff with the page
	# do I need to close the page in some way here to releae resources??

Thanks for reading

**Comment by JorjMcKie** [comment, marked as answer]
>  # do I need to close the page in some way here to releae resources??

No, this happens automatically when the name `page` is re-assigned, so no worries.
Except for the last page of course ... which will be freed by either end of script or you may assign it to None or delete it.

You are aware that your loop has a more elegant variant:
```python
for page in document:
    # process 'page'

# or from back to front:
for page in reversed(document):
    # process 'page'
```
Slice-like variants also exist.
```python
for page in document.pages(100, 13, -1):
    # ...
```

**Reply by petertennis** [reply]
Thank you for the reassurance and also the more elegant suggestion - it is most appreciated. 

I still have some performance challenges but this has allowed me to focus efforts elsewhere on other potential issues.

**Reply by petertennis** [reply]
As a follow on question are there functions that I can call on a Page object to test its integrity and especially its size? I am trying to work out why its taking so long to insert pages into my document. A potential suspect is that I have added a large number of annotations to the page - it would be good to be able to evaluate the size of the Page object before and after annotations are added. 

thanks

**Reply by JorjMcKie** [reply]
>  to test its integrity and especially its size?

You can **_clean_** the page via `page.clean_contents()`. This does what it sounds like: the potentially multiple `/Contents` objects are joined into one, display ("paint") commands in them are syntax-checked and unified, and a synchronization between this and the page's object definition takes place.

The "size" of a page can have multiple aspects:
* the amount of resources used like fonts, images, links, fields, annotations. All of this can be investigated via making respective lists and checking their lengths:

```python
annots = [a for a in page.annots()]  # the list of annots
widgets = [w for w in page.widgets()]  # fields
links = [l for l in page.links()]
fonts = page.get_fonts()
images = page.get_images()
xobjects = page.get_xobjects()
```

* The size of the (cleaned) contents simply is `len(page.read_contents())`.
* If page insertion (copy from another PDF?) is taking long, then it may mean, that the source page is huge and / or carries cross references to moe or less everything in the source document - which then of course takes a long time ...

You can insert pages from another PDF ignoring their annotations and / or links, you know.

**Reply by petertennis** [reply]
Jorj - Lots of useful suggestions to digest here - I will try them in the next few days and let you know - I really appreciate the input. 

I didn't know about the links=False option or the clean_contents option - I am going to start there and if that doesn't solve it, try to isolate the problematic pages with the other tools. 

Thanks again

**Reply by JorjMcKie** [reply]
> links=False option or the clean_contents option

oops no: this is an option for `doc.insert_pdf()`!
